fix+feat: addParticipants + createGroup

alechkos opened the pull request:
> [!IMPORTANT]
> https://github.com/pedroslopez/whatsapp-web.js/labels/breaking%20change

# Table of Contents <a href="https://github.com/pedroslopez/whatsapp-web.js/pull/2344#toc" id="toc"/>

| <h2>[`addParticipants`](https://github.com/pedroslopez/whatsapp-web.js/pull/2344#add-participants)<h3> | <h2>[`createGroup`](https://github.com/pedroslopez/whatsapp-web.js/pull/2344#create-group)<h3> |
| :---: | :---: |
| <h3>[Description](https://github.com/pedroslopez/whatsapp-web.js/pull/2344#description1)<h3> | <h3>[Description](https://github.com/pedroslopez/whatsapp-web.js/pull/2344#description2)<h3> |
| <h3>[Usage Example](https://github.com/pedroslopez/whatsapp-web.js/pull/2344#usage-example1)<h3> | <h3>[Usage Example](https://github.com/pedroslopez/whatsapp-web.js/pull/2344#usage-example2)<h3> |
| **Related Issues:**<br>closes #2154 closes #2319 closes #1943<br>closes #2067 closes #2148 closes #2229<br>closes #2158 fixes #2189 fixes #2328<br>fixes #2001 fixes #2011 | **Related Issues:**<br>closes #2243 closes #2502 fixes #2495 |

### - [I Want to Test this PR](https://github.com/pedroslopez/whatsapp-web.js/pull/2344#pr-testing)

### - [How Has the PR Been Tested (latest test on 10.10.2023)](https://github.com/pedroslopez/whatsapp-web.js/pull/2344#tests)
- [**The environment the PR was tested on**](https://github.com/pedroslopez/whatsapp-web.js/pull/2344#env)
- [**I have an issue while testing this PR**](https://github.com/pedroslopez/whatsapp-web.js/pull/2344#pr-testing-issue)

### - [Types of Changes](https://github.com/pedroslopez/whatsapp-web.js/pull/2344#types-of-changes)

---

# `addParticipants` <a href="https://github.com/pedroslopez/whatsapp-web.js/pull/2344#add-participants" id="add-participants"/>

## Description [_[Back to TOC]_](https://github.com/pedroslopez/whatsapp-web.js/pull/2344#toc) <a href="https://github.com/pedroslopez/whatsapp-web.js/pull/2344#description1" id="description1"/>
PR adds a new functionality and fixes an issue where the `addParticipants()` method would return an object with a success status even though the operation of adding participants was not successful.

Now, when trying to add participants, an object will be returned with the status of the operation for each participant. This makes it easier to understand which users were added to the group and which were not, as well as the reason behind it.
If there is an error encountered while trying to add participants, it will be returned as a string.

#### The method takes two parameters:
1. An array of participant ID's you want to add to the group: for each ID in the array, the method will attempt to add the corresponding participant to the group in the usual way. If a participant has a restriction on being added to the group (due to their privacy settings), the method will, by default, send that participant a private group invitation ([inviteV4](https://github.com/pedroslopez/whatsapp-web.js/pull/2344#invite-v4)) with an empty comment to join the group
2. Additional `options` object with properties:
    - `sleep` – the number of milliseconds to wait before adding the next participant. If it is an array, a random sleep time will be added between the `sleep[0]` and `sleep[1]` values (the difference must be at least 100 ms, otherwise, a random sleep time will be added between `sleep[1]` and `sleep[1] + 100`). If `sleep` is not an array, a sleep time equal to its value will be added. **By default, `sleep` is an array with a value of `[250, 500]`**. The purpose of that property is to avoid user's WA account from being banned, **but it might not work**. Please note that the waiting occurs only if more than one participant ID is provided
    - `autoSendInviteV4` – boolean property, if true, the inviteV4 will be sent to those participants who have restricted others from being automatically added to groups, otherwise the inviteV4 won't be sent **(true by default)**
    - `comment` – the comment to be added to an inviteV4 **(empty string by default)**, won't have an effect:
        - for all partitipant ID's in the array: if `autoSendInviteV4` is false
        - for each participant ID in the array: if the current participant does not have a restriction on being added to the group. Otherwise the comment, if provided, will be attached to all private group invitations (if any are going to be sent)

The method allows to add participants to a group even if they are not in your contact list (tested).
Additionally, if the `autoSendInviteV4` option is set to true (true by default), the inviteV4 will be sent along with a group picture thumbnail if the group has one.

<details><summary>
<strong>Types of code-message results that you could see in the returned object (click to expand):</strong>
</summary>

- 200: The participant was added successfully
- 403: The participant can be added by sending private invitation only
- 404: The phone number is not registered on WhatsApp
- 408: You cannot add this participant because they recently left the group
- 409: The participant is already a group member
- 417: The participant can't be added to the community. You can invite them privately to join this group through its invite link
- 419: The participant can't be added because the group is full
</details>

### This is how inviteV4 looks like <a href="https://github.com/pedroslopez/whatsapp-web.js/pull/2344#invite-v4" id="invite-v4"/>

- In the first example the inviteV4 belongs to the group without a profile picture
- In the second one the group has a profile picture and also the inviteV4 is sent with its thumbnail, also you can see an example when the comment is provided
![](https://github.com/pedroslopez/whatsapp-web.js/assets/93551621/dd237f8c-8110-4482-a946-7784895de09c)

## Usage Example [_[Back to TOC]_](https://github.com/pedroslopez/whatsapp-web.js/pull/2344#toc) <a href="https://github.com/pedroslopez/whatsapp-web.js/pull/2344#usage-example1" id="usage-example1"/>

```js
client.on('message', async (msg) => {
    if (msg.body === '!addmembers') {
        const group = await msg.getChat();
        const result = await group.addParticipants(['number1@c.us', 'number2@c.us', 'number3@c.us']);
        /**
         * The example of the {@link result} output:
         *
         * {
         *   'number1@c.us': {
         *     code: 200,
         *     message: 'The participant was added successfully',
         *     isInviteV4Sent: false
         *   },
         *   'number2@c.us': {
         *     code: 403,
         *     message: 'The participant can be added by sending private invitation only',
         *     isInviteV4Sent: true
         *   },
         *   'number3@c.us': {
         *     code: 404,
         *     message: 'The phone number is not registered on WhatsApp',
         *     isInviteV4Sent: false
         *   }
         * }
         */
        console.log(result);

        /**
         * If you want to provide a custom comment for private invitations
         * (the same comment will be added to all private invitations, if any are going to be sent):
         */
        await group.addParticipants(['number1@c.us', 'number2@c.us'], { comment: 'Welcome to the group' });

        /**
         * If you want to provide a special comment for each participant, you can call the method
         * with a different comment for each one (note: this will have an effect only if it is required
         * to send an inviteV4 to a participant, otherwise the participant will be added in a usual way):
         */
        await group.addParticipants(['number1@c.us'], { comment: 'Some special comment' });
        /** A single ID can also be passed as a regular string: */
        await group.addParticipants('number2@c.us', { comment: 'Some another special comment' });

        /**
         * If in the provided array there are participants
         * who have restricted others from being automatically added to groups
         * but you explicitly don't want to send them private invitations:
         */
        await group.addParticipants(['number1@c.us', 'number2@c.us'], { autoSendInviteV4: false });

        /** If you want to change the waiting value to be 100 ms: */
        await group.addParticipants(['number1@c.us', 'number2@c.us'], { sleep: 100 });

        /** If you want the method to pick the randow waiting value between 200 and 400 ms: */
        await group.addParticipants(['number1@c.us', 'number2@c.us'], { sleep: [200, 400] });
    }
});
```

---

# `createGroup` <a href="https://github.com/pedroslopez/whatsapp-web.js/pull/2344#create-group" id="create-group"/>

## Description [_[Back to TOC]_](https://github.com/pedroslopez/whatsapp-web.js/pull/2344#toc) <a href="https://github.com/pedroslopez/whatsapp-web.js/pull/2344#description2" id="description2"/>
PR extends the existing method and adds a new functionality:
- ability to create a group with youself only (where the only group creator is)
- if an array of partitipantIds is provided, the ability to automatically send [private invitations](https://github.com/pedroslopez/whatsapp-web.js/pull/2344#invite-v4) (with customizable invitation comment) to those participants whose privacy settings restrict others from adding them to groups
- ability to set a value for the message timer, after which messages will automatically disappear **(won't take an effect if the group is been creating with myself only)**
- ability to automatically link a created group to the community parent group **(won't take an effect if the group is been creating with myself only)**
- extended the returned data from the method call

Now, after the method call, an object will be returned with the group title, created group ID and the object that handles the status of the operation for each participant. This makes it easier to understand which users were added to the created group and which were not, as well as the reason behind it.
If there is an error encountered while trying to create a group, it will be returned as a string.

#### The method takes three parameters:
1. A group title
2. **(optional)** An array of participant ID's or `Contact` objects you want to add to the group.
If a participant has a restriction on being added to the group (due to their privacy settings), the method will, by default, send that participant a private group invitation ([inviteV4](https://github.com/pedroslopez/whatsapp-web.js/pull/2344#invite-v4)) with an empty comment to join the group
3. **(optional)** Object with properties:
    - `messageTimer` – the number of seconds for the messages to disappear in the group **(0 by default, won't take an effect if the group is been creating without an array of participants is provided)**
    - `parentGroupId` - the ID of a parent community group to link the newly created group with **(won't take an effect if the group is been creating without an array of participants is provided)**
    - `autoSendInviteV4` – boolean property, if true, the inviteV4 will be sent to those participants who have restricted others from being automatically added to groups, otherwise the inviteV4 won't be sent **(true by default)**
    - `comment` – the comment to be added to an inviteV4 **(empty string by default)**, won't have an effect:
        - for all partitipant ID's in the array: if `autoSendInviteV4` is false
        - for each participant ID in the array: if the current participant does not have a restriction on being added to the group. Otherwise the comment, if provided, will be attached to all private group invitations (if any are going to be sent)

## Usage Example [_[Back to TOC]_](https://github.com/pedroslopez/whatsapp-web.js/pull/2344#toc) <a href="https://github.com/pedroslopez/whatsapp-web.js/pull/2344#usage-example2" id="usage-example2"/>

```js
client.on('message', async (msg) => {
    if (msg.body === '!creategroup') {
        const partitipantsToAdd = ['number1@c.us', 'number2@c.us'];
        const result = await client.createGroup('Group Title', partitipantsToAdd);
        /**
         * The example of the {@link result} output:
         * {
         *   title: 'Group Title',
         *   gid: {
         *     server: 'g.us',
         *     user: '1111111111',
         *     _serialized: '1111111111@g.us'
         *   },
         *   participants: {
         *     'botNumber@c.us': {
         *       statusCode: 200,
         *       message: 'The participant was added successfully',
         *       isGroupCreator: true,
         *       isInviteV4Sent: false
         *     },
         *     'number1@c.us': {
         *       statusCode: 200,
         *       message: 'The participant was added successfully',
         *       isGroupCreator: false,
         *       isInviteV4Sent: false
         *     },
         *     'number2@c.us': {
         *       statusCode: 403,
         *       message: 'The participant can be added by sending private invitation only',
         *       isGroupCreator: false,
         *       isInviteV4Sent: true
         *     },
         *     'number3@c.us': {
         *       statusCode: 404,
         *       message: 'The phone number is not registered on WhatsApp',
         *       isGroupCreator: false,
         *       isInviteV4Sent: false
         *     }
         *   }
         * }
         */
        console.log(result);

        /** If you want to create a group with yourself only: */
        await client.createGroup('Group Title');

        /**
         * If in the provided array there are participants
         * who have restricted others from being automatically added to groups
         * and you explicitly don't want to send them private invitations:
         */
        await client.createGroup('Group Title', partitipantsToAdd, { autoSendInviteV4: false });

        /**
         * If you want to provide a custom comment for private invitations
         * (the same comment will be added to all private invitations, if any are going to be sent):
         */
        await client.createGroup('Group Title', partitipantsToAdd, { comment: 'Welcome to the group' });

        /** If you want to set the timer for messages (in seconds), after which messages will disappear: */
        await client.createGroup('Group Title', partitipantsToAdd, { messageTimer: 10 });

        /**
         * If you're an admin in some community and you want to create a new group and
         * automatically link it to that community:
         */
        await client.createGroup('Group Title', partitipantsToAdd, { parentGroupId: 'communityParentGroupId@g.us' });
    }
});
```

## To test this PR by yourself you can run one of the following commands: <a href="https://github.com/pedroslopez/whatsapp-web.js/pull/2344#pr-testing" id="pr-testing"/>
- **NPM**
```powershell
npm install github:alechkos/whatsapp-web.js#fix-add-participants
```
- **YARN**
```powershell
yarn add github:alechkos/whatsapp-web.js#fix-add-participants
```

## How Has This Been Tested (latest test on 10.10.2023) [_[Back to TOC]_](https://github.com/pedroslopez/whatsapp-web.js/pull/2344#toc) <a href="https://github.com/pedroslopez/whatsapp-web.js/pull/2344#tests" id="tests"/>
1. `addParticipants`
    The functionality has been tested by trying to add participants with different scenarios:
    - when the user is not yet a member of the group
    - when the user is already a member of the group
    - when the user has not restricted others from being added to groups automatically and he is not a member of the group
    - when the user has recently exited the group
    - when the user is not in the contact list
    - when the user who is attempting to add others is not an admin of the group
    - tested with trying to add a participant to the community announcement group
2. `createGroup`
    The functionality has been tested by trying to create a group in different ways:
    - when `participants` array is not provided (the group with myself only)
    - when some participants have a restriction in privacy settings to be added to groups by others
    - when participants have not restricted others from being added to groups automatically
    - providing a `messageTimer` with different values
    - when a `parentGroupId` is provided (trying to link the created group to some community parent group)

### Tested On: <a href="https://github.com/pedroslopez/whatsapp-web.js/pull/2344#env" id="env"/>

#### Types of accounts:
- Personal
- Buisness

#### Environment:
- Android 10
- Windows 10:
    - WWebJS **v1.22.2-alpha.1**
    - WWeb **v2.2333.11**, **v2.2335.9**, **v2.2338.12**, **v2.2340.9-beta**, **v2.2340.11**, **2.2340.16**, **2.2341.18**, **2.2343.55**
    - Puppeteer **v18.2.1**
    - Node **v18.17.1**
    - Chrome **v117.0.5938.92**

<details><summary>

### If you encounter any issues while testing this PR, please provide in a comment (click to expand): <a href="https://github.com/pedroslopez/whatsapp-web.js/pull/2344#pr-testing-issue" id="pr-testing-issue"/>
</summary>

1. The code you've used without any sensitive information (use [syntax highlighting](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks#syntax-highlighting) for more readability)
2. The library version
4. The WWeb version: `console.log(await client.getWWebVersion());`
5. The browser (Chrome/Chromium)
6. The error you got

> [!IMPORTANT]
> **You have to [reapply](https://github.com/pedroslopez/whatsapp-web.js/pull/2344#pr-testing) the PR each time it is changed (new commits were pushed since your last application)**
</details>

## Types of Changes [_[Back to TOC]_](https://github.com/pedroslopez/whatsapp-web.js/pull/2344#toc) <a href="https://github.com/pedroslopez/whatsapp-web.js/pull/2344#types-of-changes" id="types-of-changes"/>
- [ ] Dependency change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [X] Breaking change (fix that would cause existing functionality to change)

## Checklist
- [X] My code follows the code style of this project
- [X] I have updated the usage example accordingly (example.js)
- [X] I have updated the documentation accordingly (index.d.ts)

Commit:
added required modules

Commit:
added sendGroupV4Invite implementation

Commit:
docs updated accordingly

Commit:
bug fixed

Commit:
added usage example

Commit:
ESLint fix

Commit:
extended resulting object

Commit:
docs updated accordingly

Commit:
added required modules

Commit:
updated comments

Commit:
bug fixed

Commit:
added usage example

Commit:
fixed to catch unexpected errors

Commit:
removed unnecessary module

Commit:
updated usage example

Commit:
minor fixes

Commit:
changed to return a string if an error is occupied

Commit:
fixed usage example

tuyuribr commented:
Good work! Shouldn't we have a option to auto send the invite link if it fails (and the user isnt in the group)?

alechkos commented:
> Good work! Shouldn't we have a option to auto send the invite link if it fails (and the user isnt in the group)?

Thank you Yuri :) Yes it would be nice, but the only way is to use the implementation of #2337

alechkos commented:
What do you think if im moving the logic of [`sendGroupV4Invite()`](https://github.com/pedroslopez/whatsapp-web.js/pull/2337) to Injected.js to make it simple to be called within the `evaluate` for both methods: `sendGroupV4Invite()` and `addParticipants()` so the user will have a choice: add a participant automatically when the `addParticipants()` is called (if this option is provided), or send an invitation separately whenever he wants using the `sendGroupV4Invite()` with optional providing of invitation comment

alechkos commented:
I will close the #2337 to bring its feature here and make some changes to enable auto-sending of invitations

Commit:
Merge branch 'send-invite-v4' into fix-add-participants

alechkos commented:
Hi @Kunalvrm555 
As soon as possible

alechkos commented:
Hi @Kunalvrm555 
As soon as I have time

Commit:
removed unneccessary module

Commit:
added getAddParticipantsRpcResult

Commit:
updated addParticipants

- added AddParticipnatsOptions to provide values for sleep, autoSendInviteV4, comment
- combined addParticipants with sendGroupInviteMessage

Commit:
removed unneccessary sendGroupV4Invite

Commit:
docs updated

Commit:
fixed statement fot autoSendInviteV4

Commit:
minor fixes

- the returned value of addParticipants is no longer wrapped in an array
- docs updated
- fixed bug with the returned code value when the code is '200'

Commit:
added logic of retrieving the group profile picture

this allows to send an invitation with a profile picture thumbnail of the group

Commit:
minor fixes

Commit:
added fix for the inviteV4 to be sent

Commit:
updated usage example

Commit:
minor fix

Commit:
ESLint fix

alechkos commented:
@Kunalvrm555 
you can test it

Commit:
fix for code 419

Commit:
function name getProfilePictureThumb changed to be more clear

ronaldotantra commented:
error when running the app

node_modules/whatsapp-web.js/node_modules/puppeteer/lib/cjs/puppeteer/common/ExecutionContext.js:221
            throw new Error('Evaluation failed: ' + helper_js_1.helper.getExceptionMessage(exceptionDetails));
                  ^

Error: Evaluation failed: TypeError: Cannot read properties of undefined (reading '0')
    at window.injectToFunction (__puppeteer_evaluation_script__:122:40)
    at __puppeteer_evaluation_script__:131:12
    at ExecutionContext._evaluateInternal (/Users/ronaldotantra/Desktop/rotan/botwa/node_modules/whatsapp-web.js/node_modules/puppeteer/lib/cjs/puppeteer/common/ExecutionContext.js:221:19)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at async ExecutionContext.evaluate (/Users/ronaldotantra/Desktop/rotan/botwa/node_modules/whatsapp-web.js/node_modules/puppeteer/lib/cjs/puppeteer/common/ExecutionContext.js:110:16)
    at async Client.initialize (/Users/ronaldotantra/Desktop/rotan/botwa/node_modules/whatsapp-web.js/src/Client.js:278:9)
[nodemon] app crashed - waiting for file changes before starting...

alechkos commented:
@ronaldotantra 
> error when running the app

Most likely this is due to the fact that you are using the latest WWeb version (> 2.2327.4) and the module used in this PR was modified / removed.
In any case the PR was tested on WWeb version **2.2327.4** and it works properly.

#### To make the PR functionality work load the WWeb v2.2327.4:
- #1847 
- https://github.com/pedroslopez/whatsapp-web.js/issues/2293#issuecomment-1602462635

shirser121 commented:
I will check it in Sunday

Commit:
refactor: modules moved to window.Store.GroupParticipants

Commit:
refactor: edited link of usage examples

Commit:
refactor: LidManipulations changed to LidUtils

Commit:
Merge branch 'main' into fix-add-participants

Commit:
Merge branch 'main' into fix-add-participants

Commit:
refactor: modules regrouping

To keep the modules in order were made some changes:
- JoinInviteV4 renamed to GroupInviteV4 and grouped with module sendGroupInviteMessage
- GroupParticipants grouped with module sendAddParticipantsRPC

Commit:
fix: injectToFunction fixed to work properly

alechkos commented:
@ronaldotantra 
> error when running the app

Thank you for the note, you were right there was a small issue with `injectToFunction`
I fixed it so now you can test again :)

jrocha commented:
Testing it over here. Looks good.

tuyuribr commented on index.d.ts:
```diff
@@ -1111,58 +1111,58 @@ declare namespace WAWebJS {
-        unmute: () => Promise<void>,
+        unmute: () => Promise<void>;
         /** Returns the Contact that corresponds to this Chat. */
-        getContact: () => Promise<Contact>,
```
Why did you change this?

alechkos replied:
My fault, reverted

tuyuribr commented:
LGTM!

Commit:
revert docs changes for Chat

Commit:
Merge branch 'main' into fix-add-participants

alechkos committed:
Merge branch 'main' into fix-add-participants

diegofemello commented on index.d.ts:
```diff
@@ -1226,20 +1226,27 @@ declare namespace WAWebJS {
+        comment?: string;
+    };
+
+    export class GroupChat extends Chat {
```
this line generated errors to me 
![image](https://github.com/pedroslopez/whatsapp-web.js/assets/8761655/95b34120-353d-4839-b3eb-4537ce802aec)


changed to interface again, is solved
![image](https://github.com/pedroslopez/whatsapp-web.js/assets/8761655/32ae826d-bc84-4277-a76c-e14d75d358c6)

alechkos replied:
Thank you fot the note, fixed

Commit:
docs fixed

Commit:
minor fix in injectToFunction

Commit:
Merge branch 'main' into fix-add-participants

Commit:
Merge branch 'main' into fix-add-participants

alechkos committed:
Merge branch 'main' into fix-add-participants

gamersindo1223 commented:
Up

alechkos committed:
Merge branch 'main' into fix-add-participants

diegofemello commented:
up

alechkos committed:
Merge branch 'main' into fix-add-participants

mi-mazouz commented:
up

Commit:
Merge branch 'main' into fix-add-participants

mi-mazouz commented:
Can we merge it 🙏 ?

alechkos committed:
Merge branch 'main' into fix-add-participants

alechkos committed:
Merge branch 'main' into fix-add-participants

Commit:
removed data stringifying/parsing

Commit:
added random sleep time logic

Commit:
code refactoring

Commit:
docs updated

Gaspa79 commented:
Could we merge this pretty please? 🙏

tuyuribr commented on index.d.ts:
```diff
@@ -1229,18 +1229,25 @@ declare namespace WAWebJS {
+    };
+
+    /** AddParticipnats options */
+    export type AddParticipnatsOptions = {
```
Typo

Commit:
addParticipants method refactored

Commit:
createGroup method extended

- added functionality to send private invitations automatically to missing participants
- added ability to provide value for message timer
- added ability to link a group to the community parent group at the stage of creating a group
- extended the returned data after the method call

Commit:
docs updated + typo fixed

Commit:
added ability to create group with myself only

Commit:
example updated

Commit:
docs clarified

Commit:
small code refactoring

tuyuribr reviewed (approved):
LGTM! I'm using this in production.

mrfelipemartins commented:
@alechkos Great PR! Just ran some tests and all seems to be working fine.

Quick question, previously, calling addParticipants with a large amount of numbers was a problem, as you described, but now, while calling createGroup with a large amount of numbers wouldn't be the same? Shouldn't we have some sort of ```sleep``` when adding participants upon group creation?

alechkos commented:
@mrfelipemartins 
Thank you for testing this pr

> Shouldn't we have some sort of sleep when adding participants upon group creation?


No, as the logic for adding participants during group creation differs from the logic for simply adding participants

alechkos committed:
Merge branch 'main' into fix-add-participants

alechkos commented:
@mrfelipemartins 
Thank you for testing this PR

> Shouldn't we have some sort of sleep when adding participants upon group creation?

No, as the logic for adding participants during group creation is implemented differently (on the whatsapp side) from the logic for simply adding participants.
During the group creation, the method performs a single action by taking all the participant IDs along with other parameters (you can find implementation details in their obfuscated code).
On the other hand, the method that simply adds participants can perform the participant adding action individually for each participant in the provided array, that's why the sleep value is can be used here

mrfelipemartins commented:
I see.. thanks for clarifying. Hope this gets merged soon.

Commit:
window.Store.Invite and window.Store.InviteInfo grouped

Commit:
window.compareWwebVersions fixed

Commit:
small update to improve the performance of findImpl fix

alechkos committed:
Merge branch 'main' into fix-add-participants

alechkos committed:
Merge branch 'main' into fix-add-participants

complanboy2 commented:
While testing this PR changes, facing issue as "window.Store.GroupUtils.sendAddParticipantsRPC" as undefined...

@alechkos can you please check !

alechkos commented:
@complanboy2 
> While testing this PR changes, facing issue as "window.Store.GroupUtils.sendAddParticipantsRPC" as undefined...

I have recently tested the PR, even on the latest WWeb v2.2335.9 there are no issues with the functionality

tuyuribr commented:
createGroup don't return _serialized anymore. Its by design?

alechkos commented:
@tuyuribr
> createGroup don't return _serialized anymore. It's by design?

Yes, now the `createGroup` returns an object with group._serialized and the status for each participant that have been added during the group creation
the first participant is always the current user (who created the group)
the example of the returning object:
```js
  {
    gid: 'createdGroupId@g.us',
    participants: {
      'botNumber@c.us': {
        code: 200,
        message: 'The participant was added successfully',
        isGroupCreator: true,
        isInviteV4Sent: false
      },
      'number1@c.us': {
        code: 403,
        message: 'The participant can be added by sending private invitation only',
        isGroupCreator: false,
        isInviteV4Sent: true
      },
      'number2@c.us': {
        code: 200,
        message: 'The participant was added successfully',
        isGroupCreator: false,
        isInviteV4Sent: false
      }
    }
  }
```

tuyuribr commented:
@alechkos Shouldn't we at least keep the gid structure?
for reference without this update when creating a group it would return:

```js
"gid": {
      "server": "g.us",
      "user": "createdGroupId",
      "_serialized": "createdGroupId@g.us"
    },
    "missingParticipants": {}
```

> @tuyuribr
> 
> > createGroup don't return _serialized anymore. It's by design?
> 
> Yes, the `createGroup` returns an object with group._serialized and the status for each participant that have been added during the group creation the first participant is always the current user (who created the group) the example of the returning object:
> 
> ```js
>   {
>     gid: 'createdGroupId@g.us',
>     participants: {
>       'botNumber@c.us': {
>         code: 200,
>         message: 'The participant was added successfully',
>         isGroupCreator: true,
>         isInviteV4Sent: false
>       },
>       'number1@c.us': {
>         code: 403,
>         message: 'The participant can be added by sending private invitation only',
>         isGroupCreator: false,
>         isInviteV4Sent: true
>       },
>       'number2@c.us': {
>         code: 200,
>         message: 'The participant was added successfully',
>         isGroupCreator: false,
>         isInviteV4Sent: false
>       }
>     }
>   }
> ```

shirser121 commented:
> @alechkos Shouldn't we at least keep the gid structure?
> for reference, without this update, when we create a group it will return:
> 
> ```js
> "gid": {
>       "server": "g.us",
>       "user": "createdGroupId",
>       "_serialized": "createdGroupId@g.us"
>     },
>     "missingParticipants": {}
> ```
> I think we shouldn't break too much here, what do you think about inserting the missingParticipants again? (keeping participants and missingParticipants) So we don't break a lot 
> 
> > @tuyuribr
> > 
> > > createGroup don't return _serialized anymore. It's by design?
> > 
> > Yes, the `createGroup` returns an object with group._serialized and the status for each participant that have been added during the group creation the first participant is always the current user (who created the group) the example of the returning object:
> > 
> > ```js
> >   {
> >     gid: 'createdGroupId@g.us',
> >     participants: {
> >       'botNumber@c.us': {
> >         code: 200,
> >         message: 'The participant was added successfully',
> >         isGroupCreator: true,
> >         isInviteV4Sent: false
> >       },
> >       'number1@c.us': {
> >         code: 403,
> >         message: 'The participant can be added by sending private invitation only',
> >         isGroupCreator: false,
> >         isInviteV4Sent: true
> >       },
> >       'number2@c.us': {
> >         code: 200,
> >         message: 'The participant was added successfully',
> >         isGroupCreator: false,
> >         isInviteV4Sent: false
> >       }
> >     }
> >   }
> > ```
> 
> 

Agreed

alechkos commented:
@tuyuribr @shirser121 
I will make required changes

Commit:
gid as wid structure

Commit:
docs + example updated

alechkos committed:
Merge branch 'main' into fix-add-participants

Commit:
parenthesis

Commit:
avoid sleep after the last iteration

alechkos committed:
Merge branch 'main' into fix-add-participants

Commit:
Merge branch 'main' into fix-add-participants

ronaldotantra commented:
why i got this response but the number is not added to the group ?
```{
'6281xxx@c.us': {
    code: 200,
    message: 'The participant was added successfully',
    isInviteV4Sent: false
  }
}

alechkos commented:
@ronaldotantra 
> why i got this response but the number is not added to the group ?

The pr was tested on wweb v2.2335.9, what wweb version do you use?
`console.log(await client.getWwebVersion())`

ronaldotantra commented:
> @ronaldotantra
> 
> > why i got this response but the number is not added to the group ?
> 
> The pr was tested on wweb v2.2335.9, what wweb version do you use? `console.log(await client.getWwebVersion())`

got this error @alechkos 

console.log(await client.getWwebVersion())
                             ^

TypeError: client.getWwebVersion is not a function

alechkos commented:
@ronaldotantra 
`console.log(await client.getWWebVersion())`

ronaldotantra commented:
> @ronaldotantra `console.log(await client.getWWebVersion())`

my wweb version is 2.2338.12

alechkos commented:
@ronaldotantra 
> my wweb version is 2.2338.12

possibly that in versions >2.2335.9 wa changed something in the functionality of functions used in this pr, soon i will test it on the last wweb version but until then you can use the web cache

Commit:
fixed compatibility issue for sendGroupInviteMessage

Commit:
fixed issue with stale groupmetadata

in some cases after the participant has left the group, that participant is still stored in groupmetadata

amiruldev20 commented:
can't work
result {
  '6281358919342@c.us': {
    code: 200,
    message: 'The participant was added successfully',
    isInviteV4Sent: false
  }
}

alechkos commented:
@amiruldev20 
> can't work

i know about the issue and as i wrote in my [previous](https://github.com/pedroslopez/whatsapp-web.js/pull/2344#issuecomment-1729564279) commit i will make tests to resolve that issue
until it is fixed you can use the [web version caching](https://github.com/pedroslopez/whatsapp-web.js/pull/1847) with wweb v2.2335.9

amiruldev20 commented:
> @amiruldev20 
> > can't work
> 
> i know about the issue and as i wrote in my [previous](https://github.com/pedroslopez/whatsapp-web.js/pull/2344#issuecomment-1729564279) commit i will make tests to resolve that issue
> until it is fixed you can use the [web version caching](https://github.com/pedroslopez/whatsapp-web.js/pull/1847) with wweb v2.2335.9

Oh ok. I'll just wait for the update because I use a website that I developed without cachiny

Commit:
fixed compatibility issue for sendAddParticipantsRPC

Commit:
minor fix

alechkos commented:
@ronaldotantra @amiruldev20 
fixed

amiruldev20 commented:
> @ronaldotantra @amiruldev20 
> fixed

related to caching? I did not install caching on the web I developed like the current web. but no problem, I will try another way

alechkos commented:
@amiruldev20 
> related to caching?

no, related to the issue described [here](https://github.com/pedroslopez/whatsapp-web.js/pull/2344#issuecomment-1728671815) and [here](https://github.com/pedroslopez/whatsapp-web.js/pull/2344#issuecomment-1730991628)

amiruldev20 commented:
> @amiruldev20 
> > related to caching?
> 
> no, related to the issue described [here](https://github.com/pedroslopez/whatsapp-web.js/pull/2344#issuecomment-1728671815) and [here](https://github.com/pedroslopez/whatsapp-web.js/pull/2344#issuecomment-1730991628)

page.evaluate: TypeError: window.compareWwebVersions is not a function
 how to get this function

alechkos commented:
@amiruldev20 
> page.evaluate: TypeError: window.compareWwebVersions is not a function

You have to reapply the pr after it was updated:
```powershell
npm install github:alechkos/whatsapp-web.js#fix-add-participants
```

Commit:
jsdocs fixed

renzhxr commented:
send invite no work {
  '628@c.us': {
    code: 200,
    message: 'The participant was added successfully',
    isInviteV4Sent: false
  }
}

alechkos commented:
@renzhxr read [this](https://github.com/pedroslopez/whatsapp-web.js/pull/2344#pr-testing-issue)

ronaldotantra commented:
@alechkos got this response, but when i invite manually it works
```{
  '628@c.us': {
    code: 403,
    message: 'The participant can be added by sending private invitation only',
    isInviteV4Sent: true
  }
}

alechkos commented:
@renzhxr @ronaldotantra 

The [private invitation](https://github.com/pedroslopez/whatsapp-web.js/pull/2344#invite-v4) is sent to those participants whose privacy settings do not allow others to add them to groups
I have checked that on [several](https://github.com/pedroslopez/whatsapp-web.js/pull/2344#env) WWeb versions, including the latest one **2.2340.16**

**If you encounter any issues, you have to [reapply](https://github.com/pedroslopez/whatsapp-web.js/pull/2344#pr-testing) the PR**

renzhxr commented:
> @renzhxr @ronaldotantra 
> 
> The [private invitation](https://github.com/pedroslopez/whatsapp-web.js/pull/2344#invite-v4) is sent to those participants whose privacy settings do not allow others to add them to groups
> I have checked that on [several](https://github.com/pedroslopez/whatsapp-web.js/pull/2344#env) WWeb versions, including the latest one **2.2340.16**
> 
> **If you encounter any issues, you have to [reapply](https://github.com/pedroslopez/whatsapp-web.js/pull/2344#pr-testing) the PR**

solved thanks

Commit:
fixed issue with group creation

Commit:
fixed createGroup method

- added check for unregistered phone number
- extended returning data from method call: added title
- code refactored
- jsdocs fixed

Commit:
fixed addParticipants method

- added check for unregistered number
- code refactored

Commit:
docs fixed

Commit:
usage example fixed

Commit:
removed unnecessary method call

Commit:
jsdocs fixed

Commit:
check for group admin fixed

ronaldotantra commented:
> @renzhxr @ronaldotantra
> 
> The [private invitation](https://github.com/pedroslopez/whatsapp-web.js/pull/2344#invite-v4) is sent to those participants whose privacy settings do not allow others to add them to groups I have checked that on [several](https://github.com/pedroslopez/whatsapp-web.js/pull/2344#env) WWeb versions, including the latest one **2.2340.16**
> 
> **If you encounter any issues, you have to [reapply](https://github.com/pedroslopez/whatsapp-web.js/pull/2344#pr-testing) the PR**

if they don't allow the privacy settings to add them to groups, why i can invite them to group without sending them the private invitation ? @alechkos

alechkos commented:
@ronaldotantra 
> if they don't allow the privacy settings to add them to groups, why i can invite them to group without sending them the private invitation ?

You can't invite someone to the group without sending them the private invitation.
To invite someone literally means to send someone an invitation

ronaldotantra commented:
> @ronaldotantra
> 
> > if they don't allow the privacy settings to add them to groups, why i can invite them to group without sending them the private invitation ?
> 
> You can't invite someone to the group without sending them the private invitation. To invite someone literally means to send someone an invitation

i mean they can automatically added to group when i manually invite via phone
![CleanShot 2023-10-04 at 15 35 41](https://github.com/pedroslopez/whatsapp-web.js/assets/45096521/dc7d0eb1-b631-48e4-add1-29f5b8a6e78e)

but when i tried invite them with the code i got 403 as a response

if they do not allow the permission, i should sending invitation via personal chat like right ?
![CleanShot 2023-10-04 at 15 39 40](https://github.com/pedroslopez/whatsapp-web.js/assets/45096521/43fa4807-1630-4c5e-bdcf-72fa2f7b82dd)

so the issue is i got error 403 when they allow the permission

renzhxr commented:
> @ronaldotantra 
> > if they don't allow the privacy settings to add them to groups, why i can invite them to group without sending them the private invitation ?
> 
> You can't invite someone to the group without sending them the private invitation.
> To invite someone literally means to send someone an invitation 

bro, you seem like a professional. why don't you try making a pr combining this pr https://github.com/pedroslopez/whatsapp-web.js/pull/2509 and https://github.com/pedroslopez/whatsapp-web.js/pull/2363

alechkos commented:
@ronaldotantra 
> i mean they can automatically added to group when i manually invite via phone

If a user has disabled the option for others to add them to groups through their privacy settings, you'll need to send them a private invitation to add them to a group, whether you do it manually with your phone or programmatically with the code.

Moreover the response you got while calling the `addParticipants` method
```js
{
  'number@c.us': {
    code: 403,
    message: 'The participant can be added by sending private invitation only',
    isInviteV4Sent: true
  }
}
```
means that an error with code 403 occurred while trying to add that user to the group. This error indicates that the user has restricted others to add him to groups. Instead of adding them directly, a private invitation was sent to that user to join the group.
You can see this by the `isInviteV4Sent: true` property.

Also I just checked the method on the WWeb v2.2341.18 with the following code:
```js
client.on('message_create', async (message) => {
    if (message.fromMe && message.body.startsWith('!')) {
        const group = await message.getChat();
        console.log(await group.addParticipants(['number1@c.us', 'number2@c.us']));
    }
});
```
where `number2@c.us` can be added to groups by everyone and `number1@c.us` restricted others to add him to groups, and got a valid result as expected:
```js
{
  'number1@c.us': {
    code: 403,
    message: 'The participant can be added by sending private invitation only',
    isInviteV4Sent: true
  },
  'number2@c.us': {
    code: 200,
    message: 'The participant was added successfully',
    isInviteV4Sent: false
  }
}
```

If you encounter any issue try to [reapply](https://github.com/pedroslopez/whatsapp-web.js/pull/2344#pr-testing) the PR and also check [this](https://github.com/pedroslopez/whatsapp-web.js/pull/2344#pr-testing-issue)

shirser121 commented:
@alechkos please solve the conflicts

renzhxr commented:
Cannot add private users
{
  '15854968266@c.us': {
    code: 400,
    message: 'An unknown error occupied while adding a participant',
    isInviteV4Sent: false
  }
}

Commit:
Merge branch 'main' into fix-add-participants

alechkos commented:
@renzhxr
> Cannot add private users

Read [this](https://github.com/pedroslopez/whatsapp-web.js/pull/2344#pr-testing-issue) first
Also I just checked both methods `addParticipants` and `createGroup` on the latest WWeb v2.2343.55 and it works as expected

alechkos commented:
@shirser121 
> please solve the conflicts

Done

renzhxr commented:
Excuse me, can the member add you create in PR not run normally if the target is private (send invite). I always try to respond

{
  '68xx@c.us': {
    code: 400,
    message: 'An unknown error occupied while adding a participant',
    isInviteV4Sent: false
  }
}

I use wweb version 2.2344.52

alechkos commented:
@renzhxr 
Ah ok now I understood, you as a group member but not a group admin tries to add another mamber but get a 400 as a response
This means that the group setting `groupAdminAddMode` allows only for group admins to add new members to that group.
I will add this code to the pr implementation to clarify the response message
Thank you

renzhxr commented:
> @renzhxr 
> Ah ok now I understood, you as a group member but not a group admin tries to add another mamber but get a 400 as a response
> This means that the group setting `groupAdminAddMode` allows only for group admins to add new members to that group.
> I will add this code to the pr implementation to clarify the response message
> Thank you

bot is a group admin but the response is like that

alechkos commented:
@renzhxr 
> bot is a group admin but the response is like that

I this case I don't know why you get that response 
I tested it on several users, among them those who are in my contact list and those who are not, business accounts and regular ones, and I never got such a result

alechkos commented:
@renzhxr provide more information about your bot and members you are trying to add

renzhxr commented:
> @renzhxr provide more information about your bot and members you are trying to add

it's resolved. I tried reinstalling the modules and it was successful. Thank You

shirser121 reviewed (approved):
Thanks a lot!

kipship commented:
> Testing it over here. Looks good.

Hello how can I start

KINGMASTERAPP commented:
how to add direct group ID 
use direct group id 

 const result = await group.addParticipants(['number1@c.us', 'number2@c.us', 'number3@c.us']);

ashissharma084 commented:
can you help me to crate whatsapp and telegram group and add 200 members each group direct add? without getting banned.

RagheedBahmad commented:
Hello, I am getting the following error when trying to create a group:
 CreateGroupError: An unknown error occupied while creating a group

I made a very simple testing group and this error still haunts me. Any help would be appreciated.
